elkm1_lib/connection.py

```diff
@@ -1,287 +1,188 @@
-from typing import Optional, cast
+from collections import deque
+from functools import reduce
+from typing import Any, NamedTuple
 
+import async_timeout
 import serial_asyncio
 
 from .message import MessageEncode, decode, get_elk_command
 from .notify import Notifier
 from .util import parse_url
 
 LOG = logging.getLogger(__name__)
+HEARTBEAT_TIME = 120
+MESSAGE_RESPONSE_TIME = 5.0
+
+
+class QueuedWrite(NamedTuple):
+    """Structure for entries in the write queue."""
+
+    msg: str
+    response_cmd: str | None
+    timeout: float = 5.0
+    raw: bool = False
 
 
 class Connection:
-    """Method to manage a connection to the ElkM1."""
-
-    def __init__(
-        self,
-        url: str,
-        notifier: Notifier,
-        loop: asyncio.AbstractEventLoop | None = None,
-    ):
-        """Setup a connection."""
+    """Manage connection to ElkM1 panel."""
+
+    def __init__(self, url: str, notifier: Notifier):
         self._url = url
         self._notifier = notifier
-        self._loop = loop if loop else asyncio.get_event_loop()
 
-        self._elk_protocol: _ElkProtocol | None = None
-        self._connection_retry_time = 1
-        self._reconnect_task: asyncio.TimerHandle | None = None
+        self._writer: asyncio.StreamWriter | None = None
+        self._awaiting_response_command: str | None = None
+        self._paused = False
+        self._write_queue: deque[QueuedWrite] = deque()
+        self._check_write_queue = asyncio.Event()
+        self._response_received = asyncio.Event()
+        self._heartbeat_event = asyncio.Event()
+        self._tasks: set[asyncio.Task[Any]] = set()
 
     async def connect(self) -> None:
-        """Asyncio connection to Elk."""
+        """Create connection to Elk."""
 
         LOG.info("Connecting to ElkM1 at %s", self._url)
+        retry_time = 1
         scheme, dest, param, ssl_context = parse_url(self._url)
-        heartbeat_time = 120 if scheme != "serial" else -1
-
-        connection = partial(
-            _ElkProtocol,
-            self._loop,
-            heartbeat_time,
-            self._connected_callback,
-            self._disconnected_callback,
-            self._got_data_callback,
-            self._timeout_callback,
-        )
-        try:
+        while not self._writer:
             if scheme == "serial":
-                await serial_asyncio.create_serial_connection(
-                    self._loop, connection, dest, baudrate=param
-                )
+                coro = serial_asyncio.open_serial_connection(url=dest, baudrate=param)
             else:
-                await asyncio.wait_for(
-                    self._loop.create_connection(
-                        connection, host=dest, port=param, ssl=ssl_context
-                    ),
-                    timeout=30,
+                coro = asyncio.open_connection(host=dest, port=param, ssl=ssl_context)
+            try:
+                reader, self._writer = await asyncio.wait_for(coro, timeout=30)
```

Comment:
Usually we use `asyncio.timeout` or `async_timeout.timeout` for this pattern since it avoids creating a task

https://docs.python.org/3/library/asyncio-task.html#asyncio.timeout

```
if sys.version_info[:2] < (3, 11):
    from async_timeout import timeout as asyncio_timeout
else:
    from asyncio import timeout as asyncio_timeout
```